### DISCUSSION TITLE
batch `get` recomputations

### DISCUSSION THREAD

**Discussion by matthewhuang97** [pmndrs/jotai#1991] [Ideas] [answered]
Hi there @dai-shi, hope you're well! 

It seems that atom `get` functions aren't batched. For example:

```ts
const countAtom = atom(0);
const derivedAtom = atom((get) => {
  expensiveFn()
  return get(countAtom) + 1;
});
```

If you update `countAtom` N times in a row, we will run `expensiveFn()` N times in a row as well, if `derivedAtom` is mounted. I would've expected that we only re-compute `expensiveFn()` once async, in a batched manner. If we try reading the value of `derivedAtom` synchronously, we could lazily re-compute then if `derivedAtom` state is invalid. 

I'm not super familiar with Jotai internals, but I can imagine this would be a really significant change to fix. However, without this feature it's really tough to use Jotai as a primitive for building more data-intensive libraries! Any workarounds in the interim, would also be greatly appreciated + super helpful. Thanks so much in advance!

## Link to reproduction

https://codesandbox.io/s/peaceful-mestorf-vyvp2y

After the `setTimeout` fires, `running expensive derived getter` is printed 10 times in the console.

**Comment by dai-shi** [comment, marked as answer]
The sync behavior is by design in Jotai v2. It allows to make things behave very stably. So, I don't think we change the design.
In fact, in Jotai v1, we let React evaluate the get function, which is done lazily. So, it was batched before, but no longer.

I wonder if your use case can be solved with debounce().
One issue is we need to make it async. (We have undocumented `unstable_unwrap` to convert the async atom back to sync, if that helps.)

Here's a naive fix: https://codesandbox.io/s/mystifying-firefly-zyyrwt

Or, maybe we could create a util to wrap store for batching: `const store2 = enableBatching(store)`

**Reply by matthewhuang97** [reply]
Cool, very interesting to hear all the context! I wonder if the in-stability was due to batching in general, or batching by letting _React_ evaluate the get function? I wonder if re-introducing batching in the future, according to Jotai's own scheduler (i.e. could just be on the microtask), would be a more manageable amount of complexity / stability. 

The suggestion sounds great, and I'm excited to try it out. I did need to make one fix to `debounce`:

```ts
const debounce = (fn) => {
  let promise;
  return (...args) => {
    if (!promise) {
      promise = new Promise(async (resolve) => {
        await 0
        promise = undefined;
        resolve(fn(...args));
      });
    }
    return promise;
  };
};
```

We needed to make the promise callback run after one microtask, otherwise we would evaluate the `promise = new Promise...` setter before the `promise = undefined` line. But other than that, it looks great, and I'll try out `unstable_unwrap`.

With regards to the `enableBatching(store)` idea, do you have a rough idea of how that might be implemented?

**Reply by dai-shi** [reply]
> I wonder if the in-stability was due to batching in general

- Making updating async is troublesome, because React assumes sync.
- Sync batching is technically possible but it requires a lot of effort and React already solves it.
- v1 behavior confused many users.

**Reply by matthewhuang97** [reply]
Thanks for the details @dai-shi. Quick question, what's the difference between "unstable_unwrap" and the "loadable" util?

**Reply by dai-shi** [reply]
They are pretty similar. I considered dropping `loadable` once but we keep both because:
- the use case is a little different. `loadable` is to avoid suspense and error boundary completely for say React 17. The goal of `unwrap` isn't to avoid suspense completely and you still need error boundary.
- `loadable` has been existed since v1 and its api is rather straightforward.
- `unwrap` is more capable, like it can return the previous value.

**Comment by dai-shi** [comment]
> With regards to the `enableBatching(store)` idea, do you have a rough idea of how that might be implemented?

On second thought, my rough idea only works for primitive atoms, so it's not very useful.

```js
// a limited version: only support one argument and void result
const enableBatching = (store) => {
  const origSet = store.set;
  store.isBatching = false;
  store.pending = [];
  store.set = (atom, arg) => {
    if (!isBatching) {
      return origSet(atom, arg);
    }
    store.pending.push([atom, arg]);
  };
  store.batchedUpdate = (fn) => {
    store.isBatching = true;
    try {
      fn();
      // here comes some magic
    } finally {
      store.isBatching = false;
    }
  };
```
